filament/backend/src/metal/MetalHandles.mm

```diff
@@ -537,162 +521,242 @@ static MTLPixelFormat decidePixelFormat(id<MTLDevice> device, TextureFormat form
+    NSUInteger deviceMaxBufferLength = 256 * 1024 * 1024;   // 256 MB
     if (@available(macOS 10.14, iOS 12, *)) {
         deviceMaxBufferLength = context.device.maxBufferLength;
     }
-    if (UTILS_LIKELY(stagingBufferSize <= deviceMaxBufferLength)) {
-        auto entry = context.bufferPool->acquireBuffer(stagingBufferSize);
-        memcpy(entry->buffer.contents,
-                static_cast<uint8_t*>(data.buffer) + sourceOffset,
-                stagingBufferSize);
-        [blitCommandEncoder copyFromBuffer:entry->buffer
-                              sourceOffset:0
-                         sourceBytesPerRow:bytesPerRow
-                       sourceBytesPerImage:bytesPerSlice
-                                sourceSize:MTLSizeMake(width, height, depth)
-                                 toTexture:texture
-                          destinationSlice:slice
-                          destinationLevel:level
-                         destinationOrigin:MTLOriginMake(xoffset, yoffset, zoffset)];
-        // We must ensure we only capture a pointer to bufferPool, not "this", as this texture could
-        // be deallocated before the completion handler runs. The MetalBufferPool is guaranteed to
-        // outlive the completion handler.
-        MetalBufferPool* bufferPool = this->context.bufferPool;
-        [blitCommandBuffer addCompletedHandler:^(id<MTLCommandBuffer> cb) {
-            bufferPool->releaseBuffer(entry);
-        }];
+
+    // To upload the texture data, we either:
+    // - allocate a staging buffer and perform a buffer copy
+    // - allocate a staging texture and perform a texture blit
+    // The buffer copy is preferred, but it cannot perform format conversions or handle large uploads.
+    // The texture blit strategy does not have those limitations, however, it cannot handle 3D
+    // or cubemap texture uploads.
+
+    MTLPixelFormat stagingPixelFormat = getMetalFormat(data.format, data.type);
+    const bool conversionNecessary =
+            stagingPixelFormat != getMetalFormatLinear(devicePixelFormat) &&
+            data.type != PixelDataType::COMPRESSED;     // compressed formats should never need conversion
+
+    const bool nonBlittableTexture =
+            target == SamplerType::SAMPLER_2D_ARRAY ||
+            target == SamplerType::SAMPLER_3D ||
+            target == SamplerType::SAMPLER_CUBEMAP;
+
+    const size_t stagingBufferSize = shape.totalBytes;
+    const bool largeUpload = stagingBufferSize > deviceMaxBufferLength;
+
+    ASSERT_PRECONDITION(!nonBlittableTexture || !conversionNecessary,
+            "SAMPLER_2D_ARRAY, SAMPLER_3D, and SAMPLER_CUBEMAP texture uploads"
+            "do not support format conversions.");
+
+    ASSERT_PRECONDITION(!nonBlittableTexture || !largeUpload,
+            "SAMPLER_2D_ARRAY, SAMPLER_3D, and SAMPLER_CUBEMAP texture uploads"
+            "have a max size of %d bytes.", deviceMaxBufferLength);
```

Comment:
Is this a "todo" or just something we don't support in filament? If it's the later, we should probably have the ASSERT_PRECONDITION on the user facing API, and just use an assert() here.

Reply:
This is a limitation of the Metal backend at the moment, so more of a "todo". I'll add a TODO comment.